README.md
# tower-hyper-http-body-compat

`tower-hyper-http-body-compat` provides adapters between hyper 0.14-1.0,
http-body 0.4-1.0, and tower-service 0.3.

[![Build status](https://github.com/davidpdrsn/tower-hyper-http-body-compat/actions/workflows/CI.yml/badge.svg?branch=main)](https://github.com/davidpdrsn/tower-hyper-http-body-compat/actions/workflows/CI.yml)
[![Crates.io](https://img.shields.io/crates/v/tower-hyper-http-body-compat)](https://crates.io/crates/tower-hyper-http-body-compat)
[![Documentation](https://docs.rs/tower-hyper-http-body-compat/badge.svg)](https://docs.rs/tower-hyper-http-body-compat)

More information about this crate can be found in the [crate
documentation][docs].

## Example

Running an axum `Router` with hyper 1.0:

```rust
use axum::{Router, routing::get};
use hyper::server::conn::http1;
use std::net::SocketAddr;
use tokio::net::TcpListener;
use tower_http::trace::TraceLayer;
use tower_hyper_http_body_compat::TowerService03HttpServiceAsHyper1HttpService;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let app = Router::new()
        .route("/", get(|| async { "Hello, World!" }))
         // we can still add regular tower middleware
         .layer(TraceLayer::new_for_http());

    // `Router` implements tower-service 0.3's `Service` trait. Convert that to something
    // that implements hyper 1.0's `Service` trait.
    let service = TowerService03HttpServiceAsHyper1HttpService::new(app);

    let addr: SocketAddr = ([127, 0, 0, 1], 8080).into();

    let mut tcp_listener = TcpListener::bind(addr).await?;
    loop {
        let (tcp_stream, _) = tcp_listener.accept().await?;

        // hyper-util isn't on crates.io yet. Instead depend on it via git
        // `hyper-util = { git = "https://github.com/hyperium/hyper-util" }`
        let tcp_stream = hyper_util::rt::TokioIo::new(tcp_stream);

        let service = service.clone();
        tokio::task::spawn(async move {
            if let Err(http_err) = http1::Builder::new()
                    .keep_alive(true)
                    .serve_connection(tcp_stream, service)
                    .await {
                eprintln!("Error while serving HTTP connection: {}", http_err);
            }
        });
    }
}
```

[docs]: https://docs.rs/tower-hyper-http-body-compat


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

# Unreleased

- None.

# 0.2.0 (13. March, 2023)

- Update to hyper-1.0.0-rc.4 ([#6])

[#6]: https://github.com/davidpdrsn/tower-hyper-http-body-compat/pull/6

# 0.1.4 (13. March, 2023)

- **fixed:** Support Rust 1.60.0 ([#5])

[#5]: https://github.com/davidpdrsn/tower-hyper-http-body-compat/pull/5

# 0.1.3 (13. March, 2023)

- **added:** Implement `Default` for `HttpBody1ToHttpBody04` ([#4])

[#4]: https://github.com/davidpdrsn/tower-hyper-http-body-compat/pull/4

# 0.1.2 (26. February, 2023)

- **fixed:** Fix broken links in docs ([#3])

[#3]: https://github.com/davidpdrsn/tower-hyper-http-body-compat/pull/3

# 0.1.1 (26. February, 2023)

- **fixed:** Fix not polling again for trailers in `HttpBody1ToHttpBody04` ([#2])

[#2]: https://github.com/davidpdrsn/tower-hyper-http-body-compat/pull/2

# 0.1.0 (26. February, 2023)

- Initial release.


src/tests.rs
use std::convert::Infallible;

use bytes::Bytes;
use http::{Request, Response, StatusCode};
use http_body_util::BodyExt;
use hyper_1::server::conn::http1;
use tokio::net::TcpListener;

use crate::*;

#[tokio::test]
async fn tower_service_03_service_to_hyper_1_service() {
    async fn handle<B>(req: Request<B>) -> Result<Response<hyper_014::Body>, Infallible>
    where
        B: http_body_04::Body,
    {
        let bytes = hyper_014::body::to_bytes(req)
            .await
            .unwrap_or_else(|_| panic!());
        assert_eq!(bytes, "in");
        Ok(Response::new(hyper_014::Body::from("out")))
    }

    let svc = tower::service_fn(handle);
    let svc = TowerService03HttpServiceAsHyper1HttpService::new(svc);

    let tcp_listener = TcpListener::bind("0.0.0.0:0").await.unwrap();
    let addr = tcp_listener.local_addr().unwrap();
    tokio::task::spawn(async move {
        loop {
            let (tcp_stream, _) = tcp_listener.accept().await.unwrap();
            let tcp_stream = hyper_util::rt::TokioIo::new(tcp_stream);
            tokio::spawn(async move {
                http1::Builder::new()
                    .serve_connection(tcp_stream, svc)
                    .await
                    .unwrap();
            });
        }
    });

    let client = hyper_014::Client::builder().build_http();
    let mut res = client
        .request(
            Request::builder()
                .uri(format!("http://{addr}"))
                .body(hyper_014::Body::from("in"))
                .unwrap(),
        )
        .await
        .unwrap();

    assert_eq!(res.status(), StatusCode::OK);

    let bytes = hyper_014::body::to_bytes(&mut res).await.unwrap();
    assert_eq!(bytes, "out");
}

#[tokio::test]
async fn hyper_1_service_to_tower_service_03_service() {
    async fn handle<B>(req: Request<B>) -> Result<Response<http_body_util::Full<Bytes>>, Infallible>
    where
        B: http_body_1::Body,
    {
        let collected = req.into_body().collect().await.unwrap_or_else(|_| panic!());
        assert_eq!(collected.to_bytes(), "in");

        Ok(Response::new(http_body_util::Full::new(Bytes::from("out"))))
    }

    let svc = hyper_1::service::service_fn(handle);
    let svc = Hyper1HttpServiceAsTowerService03HttpService::new(svc);

    let tcp_listener = std::net::TcpListener::bind("0.0.0.0:0").unwrap();
    let addr = tcp_listener.local_addr().unwrap();
    tokio::spawn(async move {
        hyper_014::Server::from_tcp(tcp_listener)
            .unwrap()
            .serve(tower::make::Shared::new(svc))
            .await
            .unwrap();
    });

    let client = hyper_014::Client::builder().build_http();
    let mut res = client
        .request(
            Request::builder()
                .uri(format!("http://{addr}"))
                .body(hyper_014::Body::from("in"))
                .unwrap(),
        )
        .await
        .unwrap();

    assert_eq!(res.status(), StatusCode::OK);

    let bytes = hyper_014::body::to_bytes(&mut res).await.unwrap();
    assert_eq!(bytes, "out");
}


src/body.rs
use std::{
    pin::Pin,
    task::{Context, Poll},
};

use http::HeaderMap;
use http_body_1::Frame;
use pin_project_lite::pin_project;

// --- http-body 0.4 to http-body 1.0 ---

pin_project! {
    /// Converts an [http-body 0.4 `Body`] to an [http-body 1.0 `Body`].
    ///
    /// [http-body 0.4 `Body`]: https://docs.rs/http-body/latest/http_body/trait.Body.html
    /// [http-body 1.0 `Body`]: https://docs.rs/http-body/1.0.0-rc.2/http_body/trait.Body.html
    #[derive(Debug, Clone, Copy)]
    pub struct HttpBody04ToHttpBody1<B> {
        #[pin]
        body: B,
    }
}

impl<B> HttpBody04ToHttpBody1<B> {
    /// Create a new `HttpBody04ToHttpBody1`.
    #[inline]
    pub fn new(body: B) -> Self {
        Self { body }
    }
}

impl<B> http_body_1::Body for HttpBody04ToHttpBody1<B>
where
    B: http_body_04::Body,
{
    type Data = B::Data;
    type Error = B::Error;

    fn poll_frame(
        mut self: Pin<&mut Self>,
        cx: &mut Context<'_>,
    ) -> Poll<Option<Result<Frame<Self::Data>, Self::Error>>> {
        match self.as_mut().project().body.poll_data(cx) {
            Poll::Ready(Some(Ok(buf))) => return Poll::Ready(Some(Ok(Frame::data(buf)))),
            Poll::Ready(Some(Err(err))) => return Poll::Ready(Some(Err(err))),
            Poll::Ready(None) => {}
            Poll::Pending => return Poll::Pending,
        }

        match self.as_mut().project().body.poll_trailers(cx) {
            Poll::Ready(Ok(Some(trailers))) => Poll::Ready(Some(Ok(Frame::trailers(trailers)))),
            Poll::Ready(Ok(None)) => Poll::Ready(None),
            Poll::Ready(Err(err)) => Poll::Ready(Some(Err(err))),
            Poll::Pending => Poll::Pending,
        }
    }

    fn size_hint(&self) -> http_body_1::SizeHint {
        let size_hint = self.body.size_hint();
        let mut out = http_body_1::SizeHint::new();
        out.set_lower(size_hint.lower());
        if let Some(upper) = size_hint.upper() {
            out.set_upper(upper);
        }
        out
    }

    #[inline]
    fn is_end_stream(&self) -> bool {
        self.body.is_end_stream()
    }
}

// --- http-body 1.0 to http-body 0.4 ---

pin_project! {
    /// Converts an [http-body 1.0 `Body`] to an [http-body 0.4 `Body`].
    ///
    /// [http-body 0.4 `Body`]: https://docs.rs/http-body/latest/http_body/trait.Body.html
    /// [http-body 1.0 `Body`]: https://docs.rs/http-body/1.0.0-rc.2/http_body/trait.Body.html
    #[derive(Debug, Clone, Default)]
    pub struct HttpBody1ToHttpBody04<B> {
        #[pin]
        body: B,
        trailers: Option<HeaderMap>,
    }
}

impl<B> HttpBody1ToHttpBody04<B> {
    /// Create a new `HttpBody1ToHttpBody04`.
    #[inline]
    pub fn new(body: B) -> Self {
        Self {
            body,
            trailers: None,
        }
    }
}

impl<B> http_body_04::Body for HttpBody1ToHttpBody04<B>
where
    B: http_body_1::Body,
{
    type Data = B::Data;
    type Error = B::Error;

    fn poll_data(
        self: Pin<&mut Self>,
        cx: &mut Context<'_>,
    ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
        let this = self.project();
        match ready!(this.body.poll_frame(cx)) {
            Some(Ok(frame)) => {
                let frame = match frame.into_data() {
                    Ok(data) => return Poll::Ready(Some(Ok(data))),
                    Err(frame) => frame,
                };

                match frame.into_trailers() {
                    Ok(trailers) => {
                        *this.trailers = Some(trailers);
                    }
                    Err(_frame) => {}
                }

                Poll::Ready(None)
            }
            Some(Err(err)) => Poll::Ready(Some(Err(err))),
            None => Poll::Ready(None),
        }
    }

    fn poll_trailers(
        mut self: Pin<&mut Self>,
        cx: &mut Context<'_>,
    ) -> Poll<Result<Option<http::HeaderMap>, Self::Error>> {
        loop {
            let this = self.as_mut().project();

            if let Some(trailers) = this.trailers.take() {
                break Poll::Ready(Ok(Some(trailers)));
            }

            match ready!(this.body.poll_frame(cx)) {
                Some(Ok(frame)) => match frame.into_trailers() {
                    Ok(trailers) => break Poll::Ready(Ok(Some(trailers))),
                    // we might get a trailers frame on next poll
                    // so loop and try again
                    Err(_frame) => {}
                },
                Some(Err(err)) => break Poll::Ready(Err(err)),
                None => break Poll::Ready(Ok(None)),
            }
        }
    }

    fn size_hint(&self) -> http_body_04::SizeHint {
        let size_hint = self.body.size_hint();
        let mut out = http_body_04::SizeHint::new();
        out.set_lower(size_hint.lower());
        if let Some(upper) = size_hint.upper() {
            out.set_upper(upper);
        }
        out
    }

    #[inline]
    fn is_end_stream(&self) -> bool {
        self.body.is_end_stream()
    }
}


src/service.rs
use std::{
    fmt::Debug,
    future::Future,
    pin::Pin,
    task::{Context, Poll},
};

use pin_project_lite::pin_project;
use tower::{util::Oneshot, ServiceExt};

// --- tower-service 0.3 to hyper 1.0 ---

/// Converts a [tower-service 0.3 `Service`] to a [hyper 1.0 `Service`].
///
/// If you have a service that uses [`http::Request`] and [`http::Response`] then you probaby need
/// [`TowerService03HttpServiceAsHyper1HttpService`] instead of this.
///
/// [tower-service 0.3 `Service`]: https://docs.rs/tower-service/latest/tower_service/trait.Service.html
/// [hyper 1.0 `Service`]: https://docs.rs/hyper/1.0.0-rc.4/hyper/service/trait.Service.html
/// [`TowerService03HttpServiceAsHyper1HttpService`]: crate::TowerService03HttpServiceAsHyper1HttpService
#[derive(Clone, Copy, Debug)]
pub struct TowerService03ServiceAsHyper1Service<S>(S);

impl<S> TowerService03ServiceAsHyper1Service<S> {
    /// Create a new `TowerService03ServiceAsHyper1Service`.
    pub fn new(inner: S) -> Self {
        Self(inner)
    }
}

impl<S, R> hyper_1::service::Service<R> for TowerService03ServiceAsHyper1Service<S>
where
    S: tower_service_03::Service<R> + Clone,
{
    type Response = S::Response;
    type Error = S::Error;
    type Future = TowerService03ServiceAsHyper1ServiceFuture<S, R>;

    #[inline]
    fn call(&self, req: R) -> Self::Future {
        TowerService03ServiceAsHyper1ServiceFuture {
            // have to drive backpressure in the future
            future: self.0.clone().oneshot(req),
        }
    }
}

pin_project! {
    /// Response future for [`TowerService03ServiceAsHyper1Service`].
    pub struct TowerService03ServiceAsHyper1ServiceFuture<S, R>
    where
        S: tower_service_03::Service<R>,
    {
        #[pin]
        future: Oneshot<S, R>,
    }
}

impl<S, R> Future for TowerService03ServiceAsHyper1ServiceFuture<S, R>
where
    S: tower_service_03::Service<R>,
{
    type Output = Result<S::Response, S::Error>;

    #[inline]
    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        self.project().future.poll(cx)
    }
}

// --- hyper 1.0 to tower-service 0.3 ---

/// Converts a [hyper 1.0 `Service`] to a [tower-service 0.3 `Service`].
///
/// If you have a service that uses [`http::Request`] and [`http::Response`] then you probaby need
/// [`Hyper1HttpServiceAsTowerService03HttpService`] instead of this.
///
/// [tower-service 0.3 `Service`]: https://docs.rs/tower-service/latest/tower_service/trait.Service.html
/// [hyper 1.0 `Service`]: https://docs.rs/hyper/1.0.0-rc.4/hyper/service/trait.Service.html
/// [`Hyper1HttpServiceAsTowerService03HttpService`]: crate::Hyper1HttpServiceAsTowerService03HttpService
#[derive(Clone, Copy, Debug)]
pub struct Hyper1ServiceAsTowerService03Service<S>(S);

impl<S> Hyper1ServiceAsTowerService03Service<S> {
    /// Create a new `Hyper1ServiceAsTowerService03Service`.
    #[inline]
    pub fn new(inner: S) -> Self {
        Self(inner)
    }
}

impl<S, R> tower_service_03::Service<R> for Hyper1ServiceAsTowerService03Service<S>
where
    S: hyper_1::service::Service<R>,
{
    type Response = S::Response;
    type Error = S::Error;
    type Future = S::Future;

    #[inline]
    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        Poll::Ready(Ok(()))
    }

    #[inline]
    fn call(&mut self, req: R) -> Self::Future {
        self.0.call(req)
    }
}


src/http_service.rs
use std::{
    fmt::Debug,
    future::Future,
    marker::PhantomData,
    pin::Pin,
    task::{Context, Poll},
};

use http::{Request, Response};
use pin_project_lite::pin_project;
use tower::{util::Oneshot, ServiceExt};

use crate::{HttpBody04ToHttpBody1, HttpBody1ToHttpBody04};

// --- tower-service 0.3 (http) to hyper 1.0 (http) ---

/// Converts a [tower-service 0.3 HTTP `Service`] to a [hyper 1.0 HTTP `Service`].
///
/// An HTTP `Service` is a `Service` where the request is [`http::Request<_>`][Request] and the
/// response is [`http::Response<_>`][Response].
///
/// # Example
///
/// ```no_run
/// use http::{Request, Response, StatusCode};
/// use hyper_1::{server::conn::http1, service::service_fn, body, body::Bytes};
/// use std::{net::SocketAddr, convert::Infallible};
/// use tokio::net::TcpListener;
/// use tower_hyper_http_body_compat::TowerService03HttpServiceAsHyper1HttpService;
///
/// // a service function that uses hyper 0.14, tower-service 0.3, and http-body 0.4
/// async fn handler<B>(req: Request<B>) -> Result<Response<hyper_014::body::Body>, Infallible>
/// where
///     B: hyper_014::body::HttpBody<Data = hyper_014::body::Bytes>,
///     B::Error: Into<Box<dyn std::error::Error + Send + Sync>>,
/// {
///    let body = req.into_body();
///    let body = http_body_04::Limited::new(body, 1024);
///    let bytes = match hyper_014::body::to_bytes(body).await {
///        Ok(bytes) => bytes,
///        Err(err) => {
///            let res = Response::builder()
///                .status(StatusCode::BAD_REQUEST)
///                .body(hyper_014::body::Body::empty())
///                .unwrap();
///            return Ok(res)
///        }
///    };
///
///    let res = Response::builder()
///        .body(hyper_014::body::Body::from(format!("Received {} bytes", bytes.len())))
///        .unwrap();
///    Ok(res)
/// }
///
/// // run `handler` on hyper 1.0
/// #[tokio::main]
/// async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
///     let addr: SocketAddr = ([127, 0, 0, 1], 8080).into();
///
///     let service = tower::service_fn(handler);
///     let service = TowerService03HttpServiceAsHyper1HttpService::new(service);
///
///     let mut tcp_listener = TcpListener::bind(addr).await?;
///     loop {
///         let (tcp_stream, _) = tcp_listener.accept().await?;
///         let tcp_stream = hyper_util::rt::TokioIo::new(tcp_stream);
///         let service = service.clone();
///         tokio::task::spawn(async move {
///             if let Err(http_err) = http1::Builder::new()
///                     .keep_alive(true)
///                     .serve_connection(tcp_stream, service)
///                     .await {
///                 eprintln!("Error while serving HTTP connection: {}", http_err);
///             }
///         });
///     }
/// }
/// ```
///
/// [tower-service 0.3 HTTP `Service`]: https://docs.rs/tower-service/latest/tower_service/trait.Service.html
/// [hyper 1.0 HTTP `Service`]: https://docs.rs/hyper/1.0.0-rc.4/hyper/service/trait.Service.html
pub struct TowerService03HttpServiceAsHyper1HttpService<S, B> {
    service: S,
    _marker: PhantomData<fn() -> B>,
}

impl<S, B> TowerService03HttpServiceAsHyper1HttpService<S, B> {
    /// Create a new `TowerService03HttpServiceAsHyper1HttpService`.
    #[inline]
    pub fn new(service: S) -> Self {
        Self {
            service,
            _marker: PhantomData,
        }
    }
}

impl<S, B> Copy for TowerService03HttpServiceAsHyper1HttpService<S, B> where S: Copy {}

impl<S, B> Clone for TowerService03HttpServiceAsHyper1HttpService<S, B>
where
    S: Clone,
{
    fn clone(&self) -> Self {
        Self {
            service: self.service.clone(),
            _marker: self._marker,
        }
    }
}

impl<S, B> Debug for TowerService03HttpServiceAsHyper1HttpService<S, B>
where
    S: Debug,
{
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("TowerService03HttpServiceAsHyper1HttpService")
            .field("service", &self.service)
            .finish()
    }
}

impl<S, ReqBody, ResBody> hyper_1::service::Service<Request<ReqBody>>
    for TowerService03HttpServiceAsHyper1HttpService<S, HttpBody1ToHttpBody04<ReqBody>>
where
    S: tower_service_03::Service<
            Request<HttpBody1ToHttpBody04<ReqBody>>,
            Response = Response<ResBody>,
        > + Clone,
{
    type Response = Response<HttpBody04ToHttpBody1<ResBody>>;
    type Error = S::Error;
    type Future = TowerService03HttpServiceAsHyper1HttpServiceFuture<
        S,
        Request<HttpBody1ToHttpBody04<ReqBody>>,
    >;

    #[inline]
    fn call(&self, req: Request<ReqBody>) -> Self::Future {
        let req = req.map(HttpBody1ToHttpBody04::new);
        TowerService03HttpServiceAsHyper1HttpServiceFuture {
            future: self.service.clone().oneshot(req),
        }
    }
}

pin_project! {
    /// Response future for [`TowerService03HttpServiceAsHyper1HttpService`].
    pub struct TowerService03HttpServiceAsHyper1HttpServiceFuture<S, R>
    where
        S: tower_service_03::Service<R>,
    {
        #[pin]
        future: Oneshot<S, R>,
    }
}

impl<S, R, B> Future for TowerService03HttpServiceAsHyper1HttpServiceFuture<S, R>
where
    S: tower_service_03::Service<R, Response = Response<B>>,
{
    type Output = Result<Response<HttpBody04ToHttpBody1<B>>, S::Error>;

    #[inline]
    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        let res = ready!(self.project().future.poll(cx))?;
        Poll::Ready(Ok(res.map(HttpBody04ToHttpBody1::new)))
    }
}

// --- hyper 1.0 (http) to tower-service 0.3 (http) ---

/// Converts a [hyper 1.0 HTTP `Service`] to a [tower-service 0.3 HTTP `Service`].
///
/// An HTTP `Service` is a `Service` where the request is [`http::Request<_>`][Request] and the
/// response is [`http::Response<_>`][Response].
///
/// [tower-service 0.3 HTTP `Service`]: https://docs.rs/tower-service/latest/tower_service/trait.Service.html
/// [hyper 1.0 HTTP `Service`]: https://docs.rs/hyper/1.0.0-rc.4/hyper/service/trait.Service.html
pub struct Hyper1HttpServiceAsTowerService03HttpService<S, B> {
    service: S,
    _marker: PhantomData<fn() -> B>,
}

impl<S, B> Hyper1HttpServiceAsTowerService03HttpService<S, B> {
    /// Create a new `Hyper1HttpServiceAsTowerService03HttpService`.
    pub fn new(service: S) -> Self {
        Self {
            service,
            _marker: PhantomData,
        }
    }
}

impl<S, B> Debug for Hyper1HttpServiceAsTowerService03HttpService<S, B>
where
    S: Debug,
{
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Hyper1HttpServiceAsTowerService03HttpService")
            .field("service", &self.service)
            .finish()
    }
}

impl<S, B> Clone for Hyper1HttpServiceAsTowerService03HttpService<S, B>
where
    S: Clone,
{
    fn clone(&self) -> Self {
        Self {
            service: self.service.clone(),
            _marker: self._marker,
        }
    }
}

impl<S, B> Copy for Hyper1HttpServiceAsTowerService03HttpService<S, B> where S: Copy {}

impl<S, ReqBody, ResBody> tower_service_03::Service<Request<ReqBody>>
    for Hyper1HttpServiceAsTowerService03HttpService<S, ReqBody>
where
    S: hyper_1::service::Service<
        Request<HttpBody04ToHttpBody1<ReqBody>>,
        Response = Response<ResBody>,
    >,
{
    type Response = Response<HttpBody1ToHttpBody04<ResBody>>;
    type Error = S::Error;
    type Future = Hyper1HttpServiceAsTowerService03HttpServiceFuture<S::Future>;

    #[inline]
    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        Poll::Ready(Ok(()))
    }

    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
        let req = req.map(HttpBody04ToHttpBody1::new);
        Hyper1HttpServiceAsTowerService03HttpServiceFuture {
            future: self.service.call(req),
        }
    }
}

pin_project! {
    /// Response future for [`Hyper1HttpServiceAsTowerService03HttpService`].
    pub struct Hyper1HttpServiceAsTowerService03HttpServiceFuture<F> {
        #[pin]
        future: F,
    }
}

impl<F, B, E> Future for Hyper1HttpServiceAsTowerService03HttpServiceFuture<F>
where
    F: Future<Output = Result<Response<B>, E>>,
{
    type Output = Result<Response<HttpBody1ToHttpBody04<B>>, E>;

    #[inline]
    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        let res = ready!(self.project().future.poll(cx))?;
        Poll::Ready(Ok(res.map(HttpBody1ToHttpBody04::new)))
    }
}


src/lib.rs
//! Adapters between hyper 0.14-1.0, http-body 0.4-1.0, and tower-service 0.3.
//!
//! The required release candidates are:
//!
//! - hyper 1.0.0-rc.4
//! - http-body 1.0.0-rc.2
//!
//! # Example
//!
//! Running an axum `Router` with hyper 1.0:
//!
//! ```no_run
//! # use hyper_1 as hyper;
//! use axum::{Router, routing::get};
//! use hyper::server::conn::http1;
//! use std::net::SocketAddr;
//! use tokio::net::TcpListener;
//! use tower_http::trace::TraceLayer;
//! use tower_hyper_http_body_compat::TowerService03HttpServiceAsHyper1HttpService;
//!
//! #[tokio::main]
//! async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
//!     let app = Router::new()
//!         .route("/", get(|| async { "Hello, World!" }))
//!         // we can still add regular tower middleware
//!         .layer(TraceLayer::new_for_http());
//!
//!     // `Router` implements tower-service 0.3's `Service` trait. Convert that to something
//!     // that implements hyper 1.0's `Service` trait.
//!     let service = TowerService03HttpServiceAsHyper1HttpService::new(app);
//!
//!     let addr: SocketAddr = ([127, 0, 0, 1], 8080).into();
//!
//!     let mut tcp_listener = TcpListener::bind(addr).await?;
//!     loop {
//!         let (tcp_stream, _) = tcp_listener.accept().await?;
//!
//!         // hyper-util isn't on crates.io yet. Instead depend on it via git
//!         // `hyper-util = { git = "https://github.com/hyperium/hyper-util" }`
//!         let tcp_stream = hyper_util::rt::TokioIo::new(tcp_stream);
//!
//!         let service = service.clone();
//!         tokio::task::spawn(async move {
//!             if let Err(http_err) = http1::Builder::new()
//!                     .keep_alive(true)
//!                     .serve_connection(tcp_stream, service)
//!                     .await
//!             {
//!                 eprintln!("Error while serving HTTP connection: {}", http_err);
//!             }
//!         });
//!     }
//! }
//! ```
//!
//! Note that this library doesn't require axum. Its supports any [`tower::Service`].
//!
//! # Feature flags
//!
//! To enable the `Service` adapters you must enable either `http1` or `http2` and `server` or
//! `client` (i.e. `(http1 || http2) && (server || client)`).
//!
//! The `Body` adapters are always enabled.

#![warn(
    clippy::all,
    clippy::dbg_macro,
    clippy::todo,
    clippy::empty_enum,
    clippy::enum_glob_use,
    clippy::mem_forget,
    clippy::unused_self,
    clippy::filter_map_next,
    clippy::needless_continue,
    clippy::needless_borrow,
    clippy::match_wildcard_for_single_variants,
    clippy::if_let_mutex,
    clippy::mismatched_target_os,
    clippy::await_holding_lock,
    clippy::match_on_vec_items,
    clippy::imprecise_flops,
    clippy::suboptimal_flops,
    clippy::lossy_float_literal,
    clippy::rest_pat_in_fully_bound_structs,
    clippy::fn_params_excessive_bools,
    clippy::exit,
    clippy::inefficient_to_string,
    clippy::linkedlist,
    clippy::macro_use_imports,
    clippy::option_option,
    clippy::verbose_file_reads,
    clippy::unnested_or_patterns,
    clippy::str_to_string,
    rust_2018_idioms,
    future_incompatible,
    nonstandard_style,
    missing_debug_implementations,
    missing_docs
)]
#![deny(unreachable_pub, private_in_public)]
#![allow(elided_lifetimes_in_paths, clippy::type_complexity)]
#![forbid(unsafe_code)]
#![cfg_attr(docsrs, feature(doc_auto_cfg, doc_cfg))]
#![cfg_attr(test, allow(clippy::float_cmp))]

macro_rules! cfg_service {
    ($($item:item)*) => {
        $(
            #[cfg(all(
                any(feature = "http1", feature = "http2"),
                any(feature = "server", feature = "client")
            ))]
            $item
        )*
    };
}

macro_rules! ready {
    ($e:expr) => {
        match $e {
            std::task::Poll::Ready(t) => t,
            std::task::Poll::Pending => {
                return std::task::Poll::Pending;
            }
        }
    };
}

cfg_service! {
    mod service;
    mod http_service;

    pub use service::{Hyper1ServiceAsTowerService03Service, TowerService03ServiceAsHyper1Service};
    pub use http_service::{
        Hyper1HttpServiceAsTowerService03HttpService, TowerService03HttpServiceAsHyper1HttpService,
    };
}

mod body;

pub use body::{HttpBody04ToHttpBody1, HttpBody1ToHttpBody04};

#[cfg(test)]
mod tests;

pub mod future {
    //! Future types.

    cfg_service! {
        pub use crate::http_service::{
            Hyper1HttpServiceAsTowerService03HttpServiceFuture,
            TowerService03HttpServiceAsHyper1HttpServiceFuture,
        };
        pub use crate::service::TowerService03ServiceAsHyper1ServiceFuture;
    }
}
